opnieuw/retries.py

```diff
@@ -68,98 +68,84 @@ def calculate_exponential_multiplier(
-    # Number of seconds left before the user's requested seconds are exceeded
-    seconds_left: float
-
-
-Action = Union[DoCall, DoWait]
-
-
-class RetryState:
-    """
-    Calculate whether to do the function call or to sleep for retrying.
+    Will consider the maximum amount of backoffs and a maximum backoff window.
     """
 
     def __init__(
         self,
         clock: Clock,
-        max_calls_total: int,
-        retry_window_after_first_call_in_seconds: int,
+        max_backoffs_total: int,
+        backoff_window_after_first_call_in_seconds: int,
     ) -> None:
         self.clock = clock
-        self.max_calls_total = max_calls_total
+        self.max_backoffs_total = max_backoffs_total
         self.deadline_second = (
-            self.clock.seconds_since_epoch() + retry_window_after_first_call_in_seconds
+            self.clock.seconds_since_epoch()
+            + backoff_window_after_first_call_in_seconds
         )
-
         self.base_in_seconds = calculate_exponential_multiplier(
-            max_calls_total, retry_window_after_first_call_in_seconds
+            max_backoffs_total, backoff_window_after_first_call_in_seconds
         )
+        self.backoffs = 0
 
-    def __iter__(self) -> Iterator[Action]:
-        for attempt in range(0, self.max_calls_total):
-            # attempt the actual function call
-            yield DoCall()
+    def get_backoff(self) -> float | None:
+        """
+        Return the amount of seconds to backoff.
 
-            wait_seconds = self.base_in_seconds * 2**attempt
-            seconds_left = self.deadline_second - self.clock.seconds_since_epoch()
+        None indicates that there should be no more backoffs.
+        """
+        backoff_seconds = self.base_in_seconds * 2**self.backoffs
+        jittered_backoff = random.uniform(0.0, backoff_seconds)
 
-            # signal that we need to sleep
-            yield DoWait(
-                attempts_so_far=attempt + 1,
-                min_seconds=0.0,
-                max_seconds=wait_seconds,
-                seconds_left=seconds_left,
-            )
+        self.backoffs += 1
+        if self.backoffs >= self.max_backoffs_total:
+            logger.debug(f"Used up all {self.backoffs} retries.")
+            return None
 
+        if jittered_backoff > self.deadline_second - self.clock.seconds_since_epoch():
+            logger.debug("Next attempt would be after retry deadline, not retrying.")
+            return None
 
-__retry_state_namespaces: dict[str | None, ContextVar[type[RetryState]]] = defaultdict(
-    lambda: ContextVar("opnieuw_default_retry_state", default=RetryState)
+        logger.debug(
+            f"Backoff for {jittered_backoff:.3f} seconds after attempt {self.backoffs}"
+        )
+        return jittered_backoff
+
+
+__backoff_namespaces: dict[
+    str | None, ContextVar[type[BackoffCalculator]]
+] = defaultdict(
+    lambda: ContextVar("opnieuw_default_retry_state", default=BackoffCalculator)
```

Comment:
This can be renamed too

```suggestion
    lambda: ContextVar("opnieuw_default_backoff_state", default=BackoffCalculator)
```